File: src/main/java/data/stack/SimplifyPath_71.java
package data.stack;

import java.util.Arrays;
import java.util.Stack;

public class SimplifyPath_71 {
  public static String simplifyPath(String path) {
    var stack = new Stack<String>();
    Arrays.stream(path.split("/")).filter(a -> !a.isBlank()).forEach(p -> {
      switch (p) {
        case ".." -> {
          if (!stack.isEmpty()) {
            stack.pop();
          }
        }
        case "." -> {}
        default -> stack.push(p);
      }
    });

    return "/" + String.join("/", stack);
  }

  public static void main(String[] args) {
    System.out.println(simplifyPath("/home"));
    System.out.println(simplifyPath("/../"));
    System.out.println(simplifyPath("/home//foo/"));
    System.out.println(simplifyPath("/a/./b/../../c/"));
    System.out.println(simplifyPath("/a/../../b/../c//.//"));
    System.out.println(simplifyPath("/a//b////c/d//././.."));
  }
}

File: src/main/kotlin/data/structure/stack/SimplifyPath_71.kt
package data.structure.stack

import java.util.Stack

class KtSimplifyPath_71 {
    fun simplifyPath(path: String): String {
        val stack = Stack<String>()
        path.split("/").filter { it.isNotBlank() }.forEach { folder ->
            when (folder) {
                ".." -> {
                    if (stack.isNotEmpty()) {
                        stack.pop()
                    }
                }
                "." -> { /* do nothing */ }
                else -> {
                    stack.push(folder)
                }
            }
        }

        return "/" + stack.joinToString("/")
    }
}

fun main() {
    val simplify = KtSimplifyPath_71()
    println(simplify.simplifyPath("/a/./b/../../c/"))
}